File: mobile/lib/shared/services/api/payment_service.dart
import '../api/api_client.dart';
import '../api/api_config.dart';
import 'package:meta/meta.dart';

part 'payment_service.g.dart';

@JsonSerializable()
class Payment {
  final String id;
  final String userId;
  final double amount;
  final String currency;
  final String status;
  final String method;
  final String reference;
  final String? description;
  final Map<String, dynamic>? metadata;
  final String createdAt;
  final String updatedAt;

  const Payment({
    required this.id,
    required this.userId,
    required this.amount,
    required this.currency,
    required this.status,
    required this.method,
    required this.reference,
    this.description,
    this.metadata,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Payment.fromJson(Map<String, dynamic> json) => _$PaymentFromJson(json);
  Map<String, dynamic> toJson() => _$PaymentToJson(this);
}

@JsonSerializable()
class PaymentGateway {
  final String id;
  final String name;
  final String code;
  final bool isActive;
  final Map<String, dynamic> config;

  const PaymentGateway({
    required this.id,
    required this.name,
    required this.code,
    required this.isActive,
    required this.config,
  });

  factory PaymentGateway.fromJson(Map<String, dynamic> json) => _$PaymentGatewayFromJson(json);
  Map<String, dynamic> toJson() => _$PaymentGatewayToJson(this);
}

@JsonSerializable()
class Wallet {
  final String id;
  final String userId;
  final double balance;
  final String currency;
  final String createdAt;
  final String updatedAt;

  const Wallet({
    required this.id,
    required this.userId,
    required this.balance,
    required this.currency,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Wallet.fromJson(Map<String, dynamic> json) => _$WalletFromJson(json);
  Map<String, dynamic> toJson() => _$WalletToJson(this);
}

@JsonSerializable()
class WalletTransaction {
  final String id;
  final String walletId;
  final String type;
  final double amount;
  final String description;
  final String? reference;
  final double balanceAfter;
  final String createdAt;

  const WalletTransaction({
    required this.id,
    required this.walletId,
    required this.type,
    required this.amount,
    required this.description,
    this.reference,
    required this.balanceAfter,
    required this.createdAt,
  });

  factory WalletTransaction.fromJson(Map<String, dynamic> json) => _$WalletTransactionFromJson(json);
  Map<String, dynamic> toJson() => _$WalletTransactionToJson(this);
}

class PaymentService {
  final ApiClient _client;

  PaymentService({ApiClient? client}) : _client = client ?? ApiClient();

  // List payments
  Future<ApiResponse<List<Payment>>> listPayments({
    int page = 1,
    int limit = 20,
    String? status,
    String? userId,
    String? startDate,
    String? endDate,
  }) async {
    final queryParams = <String, dynamic>{};
    queryParams['page'] = page;
    queryParams['limit'] = limit;
    if (status != null) queryParams['status'] = status;
    if (userId != null) queryParams['userId'] = userId;
    if (startDate != null) queryParams['startDate'] = startDate;
    if (endDate != null) queryParams['endDate'] = endDate;

    return _client.get<List<Payment>>(
      '/payments',
      queryParams: queryParams,
      fromJson: (json) {
        final data = json['data'] as List<dynamic>? ?? [];
        return data.map((item) => Payment.fromJson(item as Map<String, dynamic>)).toList();
      },
    );
  }

  // Get payment by ID
  Future<ApiResponse<Payment>> getPayment(String paymentId) async {
    return _client.get<Payment>(
      '/payments/$paymentId',
      fromJson: Payment.fromJson,
    );
  }

  // Create payment
  Future<ApiResponse<Payment>> createPayment({
    required double amount,
    String currency = 'NGN',
    required String method,
    String? reference,
    String? description,
    Map<String, dynamic>? metadata,
  }) async {
    return _client.post<Payment>(
      '/payments',
      body: {
        'amount': amount,
        'currency': currency,
        'method': method,
        'reference': reference,
        'description': description,
        'metadata': metadata,
      },
      fromJson: Payment.fromJson,
    );
  }

  // Verify payment
  Future<ApiResponse<Map<String, dynamic>>> verifyPayment(String reference) async {
    return _client.post<Map<String, dynamic>>(
      '/payments/verify',
      body: {'reference': reference},
    );
  }

  // List payment gateways
  Future<ApiResponse<List<PaymentGateway>>> listPaymentGateways() async {
    return _client.get<List<PaymentGateway>>(
      '/payments/gateways',
      fromJson: (json) {
        final data = json['data'] as List<dynamic>? ?? [];
        return data.map((item) => PaymentGateway.fromJson(item as Map<String, dynamic>)).toList();
      },
    );
  }

  // Get my wallet
  Future<ApiResponse<Wallet>> getMyWallet() async {
    return _client.get<Wallet>(
      '/payments/wallet',
      fromJson: Wallet.fromJson,
    );
  }

  // List wallet transactions
  Future<ApiResponse<List<WalletTransaction>>> listWalletTransactions({
    int page = 1,
    int limit = 20,
  }) async {
    return _client.get<List<WalletTransaction>>(
      '/payments/wallet/transactions',
      queryParams: {'page': page, 'limit': limit},
      fromJson: (json) {
        final data = json['data'] as List<dynamic>? ?? [];
        return data.map((item) => WalletTransaction.fromJson(item as Map<String, dynamic>)).toList();
      },
    );
  }

  // Fund wallet
  Future<ApiResponse<Map<String, dynamic>>> fundWallet({
    required double amount,
    required String paymentMethodId,
  }) async {
    return _client.post<Map<String, dynamic>>(
      '/payments/wallet/fund',
      body: {
        'amount': amount,
        'paymentMethodId': paymentMethodId,
      },
    );
  }
}

File: web/src/services/api/paymentService.ts
import { getSupabase } from '@/lib/supabase';
import type { PaginatedResponse } from '@/types/api/api';

export interface Payment { id:string; userId:string; amount:number; currency:string; status:'pending'|'completed'|'failed'|'refunded'; method?:string; gateway?:string; reference:string; description?:string; metadata?:Record<string,unknown>; createdAt:string; updatedAt:string; }
export interface PaymentFilters { page?:number; limit?:number; status?:string; userId?:string; startDate?:string; endDate?:string; }
export interface CreatePaymentData { amount:number; currency?:string; gateway:'paystack'|'flutterwave'; planId?:string; courseId?:string; examId?:string; redirectUrl?:string; metadata?:Record<string,unknown>; }
export interface CreatePaymentResponse { success:boolean; message?:string; data:{ payment:Payment; authorizationUrl?:string|null; accessCode?:string|null; reference?:string; }; }
export interface PaymentGateway { id:string; name:string; code:string; isActive:boolean; config?:Record<string,unknown>; }

const mapPayment=(row:any):Payment=>({id:row.id,userId:row.user_id,amount:Number(row.amount||0),currency:row.currency,status:row.status,gateway:row.gateway,reference:row.reference,metadata:row.metadata||{},createdAt:row.created_at,updatedAt:row.updated_at});
const invoke=async<T>(body:Record<string,unknown>):Promise<T>=>{const {data,error}=await getSupabase().functions.invoke('payments',{body});if(error)throw new Error(error.message||'Payment operation failed');if(data?.error)throw new Error(String(data.error));return data as T;};

export const fetchPayments=async(filters:PaymentFilters={},_token:string):Promise<PaginatedResponse<Payment>>=>{const user=(await getSupabase().auth.getUser()).data.user;if(!user)throw new Error('You must be signed in');const page=Math.max(1,filters.page||1);const limit=Math.min(100,Math.max(1,filters.limit||20));const from=(page-1)*limit;let query=getSupabase().from('payments').select('*',{count:'exact'}).eq('user_id',user.id).order('created_at',{ascending:false}).range(from,from+limit-1);if(filters.status)query=query.eq('status',filters.status);const {data,error,count}=await query;if(error)throw new Error(error.message);return{data:(data||[]).map(mapPayment),page,pageSize:limit,total:count||0,totalPages:Math.ceil((count||0)/limit)};};
export const fetchPaymentById=async(paymentId:string,_token:string):Promise<{payment:Payment}>=>{const user=(await getSupabase().auth.getUser()).data.user;if(!user)throw new Error('You must be signed in');const {data,error}=await getSupabase().from('payments').select('*').eq('id',paymentId).eq('user_id',user.id).maybeSingle();if(error||!data)throw new Error(error?.message||'Payment not found');return{payment:mapPayment(data)};};
export const createPayment=async(data:CreatePaymentData,_token:string):Promise<CreatePaymentResponse>=>invoke<CreatePaymentResponse>({action:'create-payment',...data});
export const verifyPayment=async(reference:string,_token:string,transactionId?:string)=>invoke({action:'verify-payment',reference,transactionId});
export const fetchPaymentGateways=async(_token?:string):Promise<{success?:boolean;data?:{gateways:PaymentGateway[]};gateways?:PaymentGateway[]}>=>{const result=await invoke<{gateways:PaymentGateway[]}>({action:'get-gateways'});return{gateways:result.gateways,data:{gateways:result.gateways}};};

export interface Wallet { id:string; userId:string; balance:number; currency:string; createdAt:string; updatedAt:string; }
export interface WalletTransaction { id:string; walletId:string; type:'credit'|'debit'; amount:number; description:string; reference?:string; balanceAfter:number; createdAt:string; }
export const fetchMyWallet=async(_token:string):Promise<{wallet:Wallet}>=>{throw new Error('Wallet funding is not enabled in this release.');};
export const fetchWalletTransactions=async(_page=1,_limit=20,_token:string):Promise<PaginatedResponse<WalletTransaction>>=>({data:[],page:_page,pageSize:_limit,total:0,totalPages:0});
export const fundWallet=async(_amount:number,_paymentMethodId:string,_token:string)=>{throw new Error('Wallet funding is not enabled in this release.');};
